src/Matrix.jl

```diff
@@ -3686,6 +3686,33 @@ function nullspace(M::MatElem{T}) where {T <: FieldElement}
    return nullity, X
 end
 
+###############################################################################
+#
+#   Nilpotency
+#
+###############################################################################
+
+@doc raw"""
+    is_nilpotent(A::MatrixElem{T}) where {T <: RingElement}
+
+Return if `A` is nilpotent, i.e. if there exists a natural number $k$
+such that $A^k = 0$. If `A` is not square an exception is raised.
+"""
+function is_nilpotent(A::MatrixElem{T}) where {T <: RingElement}
+  !is_square(A) && error("Dimensions don't match in is_nilpotent")
+  is_zero(tr(A)) || return false
+  n = nrows(A)
+  A = deepcopy(A)
+  i = 1
+  is_zero(A) && return true
+  while i < n
```

Comment:
Actually, is this algorithm even correct when `A` contains zero divisors? I normally think about matrices over integral domains, and then of course `A` is nilpotent iff `A^n=0`. But over a ring with zero-divisors, we could have a triangularizable matrix with nilpotent eigenvalues where $A^n\neq0$ but still $A^k=0$ for some $k>n$.

Admittedly this is somewhat fringe, but we should still handle it right? For starters we could add `is_domain_type(T) || error("Only supported over integral domains")` ? (AFAIK there is no `is_domain` / `is_integral_domain` which for e.g. $\mathbb{Z}/n\mathbb{Z}$ takes the modulus $n$ into account)

Then if someone needs this over non-domains, they at least get a helpful error and can provide an implementation for their use case.

Reply:
Good catch! Thanks for mentioning this. I added the suggested error